File: hw/fuentes/U3/HW21TEMPLATE/HW21TEMPLATEPY/ec/edu/espe/cafe/model/CaffeineBeverage.py
class CaffeineBeverage:
    def prepareRecipe(self):
        self.boilWater()
        self.brew()
        self.pourInCup()
        if self.wantsCondiments():
            self.addCondiments()

    def boilWater(self):
        print("Boiling water")

    def pourInCup(self):
        print("Pouring into cup")

    def brew(self):
        raise NotImplementedError("Subclasses must implement brew()")

    def addCondiments(self):
        raise NotImplementedError("Subclasses must implement addCondiments()")

    def wantsCondiments(self):
        return True

File: hw/vaca/u3/HWTemplateMethod/mavenproject1/src/main/java/ec/edu/espe/mavenproject1/CaffeineBeverage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 */

package ec.edu.espe.mavenproject1;

/**
 *
 * @author Kevin Vaca Edison's OOP ESPE
 */
public abstract class CaffeineBeverage {

public final void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) {
            addCondiments();
        }
    }
    
    // Common implementation for all beverages
    private void boilWater() {
        System.out.println("Boiling water");
    }
    
    // Common implementation for all beverages
    private void pourInCup() {
        System.out.println("Pouring into cup");
    }
    
    // Primitive operations to be implemented by subclasses
    protected abstract void brew();
    protected abstract void addCondiments();
    
    // Hook - subclasses can override if needed
    protected boolean wantsCondiments() {
        return true;
    }
}
